### PR TITLE
Prevent consecutive ::cs_main locks

### PR DESCRIPTION
**Pull request by hebasto**
This PR prevents consecutive `::cs_main` locks in the following functions:
- `FlushStateToDisk()` that could happen when it is called from:
  - `AcceptToMemoryPoolWithTime()`
  - `CChainState::DisconnectTip()`
  - `CChainState::ConnectTip()`
  - `CChainState::AcceptBlock()`
  - `CChainState::RewindBlockIndex()`
- UnloadBlockIndex()
- CChainState::ReplayBlocks()

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2234,7 +2234,6 @@ bool CChainState::FlushStateToDisk(
     FlushStateMode mode,
     int nManualPruneHeight)
 {
-    LOCK(cs_main);
```
Comment (MarcoFalke) [inline src/validation.cpp:2237]: Removed locks are generally replaced with `AssertLockHeld(cs_main);` and the corresponding compiler annotations

**Comment by hebasto** [comment]
Updated 9b4b3c9442f5b2ccdce8d70161a355aec5402cef  -> fe187562ef9765942d7c7d801e04d70d294d5aef ([pr18639.01](https://github.com/hebasto/bitcoin/commits/pr18639.01) -> [pr18639.02](https://github.com/hebasto/bitcoin/commits/pr18639.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18639.01..pr18639.02)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18639#discussion_r408447328):
> Removed locks are generally replaced with `AssertLockHeld(cs_main);` and the corresponding compiler annotations

File: src/validation.cpp (hunk 1 above)
Comment (hebasto) [reply src/validation.cpp:2237]: [Updated](https://github.com/bitcoin/bitcoin/pull/18639#issuecomment-613708604).

Comment (MarcoFalke) [reply src/validation.cpp:2237]: Please also add the annotation in the header

Comment (hebasto) [reply src/validation.cpp:2237]: > Please also add the annotation in the header
> 
> EXCLUSIVE_LOCKS_REQUIRED(cs_main)

It is already done: https://github.com/bitcoin/bitcoin/blob/fe187562ef9765942d7c7d801e04d70d294d5aef/src/validation.h#L671-L675

**Comment by hebasto** [comment]
Updated fe187562ef9765942d7c7d801e04d70d294d5aef  -> 18f0cb80c0cea7c9bc0b69a0f1fe0ca6cae8cde1 ([pr18639.02](https://github.com/hebasto/bitcoin/commits/pr18639.02) -> [pr18639.03](https://github.com/hebasto/bitcoin/commits/pr18639.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18639.02..pr18639.03)):

analogous changes added for:
- `UnloadBlockIndex()`
- `CChainState::ReplayBlocks()`

**Comment by MarcoFalke** [comment]
What problem is this solving?

**Comment by hebasto** [comment]
@MarcoFalke 
> What problem is this solving?

Currently, `::cs_main` has type `RecursiveMutex` which derives from `std::recursive_mutex`.

But I'm agree with the following `TODO` statement:
https://github.com/bitcoin/bitcoin/blob/4a71c469058b824ed6c5132ab9901e194fa8aae1/src/sync.h#L106-L110

**Comment by MarcoFalke** [comment]
Ok, Concept ACK based on that motivation, but OP should probably mention that. Some general notes, though:

* I am not sure if we can ever make `cs_main` not recursive. There might be circular dependencies in similar ways to how `csPathCached` can not be made a Mutex right now.
* We currently have no mechanism to prevent double locking a Mutex. This is undefined behavior and currently happens on some of the Mutexes we have. Someone should probably fix that.

**Comment by hebasto** [comment]
>* We currently have no mechanism to prevent double locking a Mutex. This is undefined behavior and currently happens on some of the Mutexes we have. Someone should probably fix that.

Thanks for reminding :)

**Comment by MarcoFalke** [comment]
tsan should detect those: https://github.com/llvm/llvm-project/blame/master/compiler-rt/test/tsan/mutex_double_lock.cpp

**Comment by promag** [comment]
Code review ACK e09b98605f6fdd2ad3bcdf4c68d229905f7a29cd.

I don't see much value in this change especially because a) it's touching validation b) it's not fixing a problem.

**Pull request closed without merging**